src/events.rs

```diff
@@ -0,0 +1,296 @@
+        }))
+        .await
+    }
+
+    pub async fn worker_start_failed(&mut self, runtime_name: &str, err: Arc<worker::StartError>) {
+        self.notify(Event::WorkerStartFailed(
+            NodeData {
+                runtime_name: runtime_name.to_owned(),
+            },
+            err,
+        ))
+        .await
+    }
+
+    pub async fn worker_terminated(&mut self, runtime_name: &str) {
+        self.notify(Event::WorkerTerminated(NodeData {
+            runtime_name: runtime_name.to_owned(),
+        }))
+        .await
+    }
+
+    pub async fn worker_termination_failed(
+        &mut self,
+        runtime_name: &str,
+        err: Arc<worker::TerminationError>,
+    ) {
+        self.notify(Event::WorkerTerminationFailed(
+            NodeData {
+                runtime_name: runtime_name.to_owned(),
+            },
+            err,
+        ))
+        .await
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// EventBufferCollector is an event listener that collects all the events that
+/// have been published by a supervision tree. It later can be used to assert
+/// events that have happened.
+pub struct EventBufferCollector {
+    events: Arc<Mutex<Vec<Event>>>,
+    join_handle: JoinHandle<()>,
+}
+
+impl EventBufferCollector {
+    pub async fn from_mpsc(receiver: mpsc::Receiver<Event>) -> EventBufferCollector {
+        let events = Arc::new(Mutex::new(Vec::new()));
+        let join_handle = task::spawn(run_event_collector(events.clone(), receiver));
+        EventBufferCollector {
+            events,
+            join_handle,
+        }
+    }
+
+    /// get_events returned the number of events that have happened so far
+    pub async fn get_events(&self) -> Vec<Event> {
+        let events = self.events.lock().await;
+        (*events).clone()
+    }
+
+    /// assert_exact checks that the accumulated events that have happened so
+    /// far match the given assertions in order.
+    pub async fn assert_exact(&self, asserts: Vec<EventAssert>) {
+        let events = self.get_events().await;
+        assert_eq!(events.len(), asserts.len(), "{:?}", events);
+        for (ev, assert) in events.into_iter().zip(asserts.into_iter()) {
+            assert.check(ev)
+        }
+    }
+
+    // TODO: A function that will block current thread until an event assert is
+    // successful
+    //
+    // pub async fn wait_till(&self, assert: EventAssert, timeout: time::Duration)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// EventAssert is a well-defined function that asserts properties from an Event
+/// emitted by a running supervision tree.
+pub struct EventAssert(Box<dyn Fn(Event) -> String>);
+
+impl EventAssert {
+    fn call(&self, ev: Event) -> String {
+        (*self.0)(ev)
+    }
+    pub fn check(&self, ev: Event) {
+        let result = self.call(ev);
+        if result.len() != 0 {
+            assert!(false, result);
```

Comment:
Since this is basically a way to do a guaranteed `panic!` I think you should just use `panic!` directly. You can format your own error using format syntax. Something like:

```suggestion
            panic!("EventAssert failed: {}", result);
```